File: python/searchable-client/src/searchable_client/score.py
import math

from searchable_client.types import Manifest, Posting

K1 = 1.2
B = 0.75


def _idf(doc_count: int, df: int) -> float:
    return math.log(1 + (doc_count - df + 0.5) / (df + 0.5))


def score_term_for_doc(
    posting: Posting,
    df: int,
    manifest: Manifest,
    language: str,
    field_boost_overrides: dict[str, float] | None = None,
) -> float:
    avg_field_length = manifest.avg_field_length.get(language, {})
    weighted_tf = 0.0
    for field_name, field_posting in posting.fields.items():
        boost = (field_boost_overrides or {}).get(field_name) or (
            manifest.fields[field_name].boost if field_name in manifest.fields else 1.0
        )
        avg_len = avg_field_length.get(field_name) or field_posting.len
        length_norm = 1 - B + B * (field_posting.len / (avg_len or 1))
        weighted_tf += (boost * field_posting.tf) / (length_norm or 1)

    doc_count = manifest.doc_count.get(language, 0)
    return _idf(doc_count, df) * (weighted_tf / (weighted_tf + K1))

File: packages/searchable/src/score.ts
import { ownProp } from "./analysis/index.js";
import type { Manifest, Posting } from "./format/index.js";

/** BM25F parameters (docs/guides/ranking-and-boosts.md#ranking-model-bm25f). */
const K1 = 1.2;
const B = 0.75;

function idf(docCount: number, df: number): number {
  return Math.log(1 + (docCount - df + 0.5) / (df + 0.5));
}

/**
 * Scores one term's contribution to one document under BM25F: field-
 * weighted term frequency (normalized by that field's length relative
 * to the corpus average), run through the BM25 saturation curve and
 * scaled by idf. `fieldBoostOverrides` lets a caller override the
 * manifest's build-time field weights per-query
 * (docs/guides/ranking-and-boosts.md#boost-types-summarized). `language`
 * selects which partition's corpus-wide stats (docCount, avgFieldLength)
 * to use — a manifest spanning multiple languages keeps these per
 * language, since mixing them would skew idf/length-norm for both
 * (docs/guides/internationalization.md#mixed-language-corpora-and-queries).
 */
export function scoreTermForDoc(
  posting: Posting,
  df: number,
  manifest: Manifest,
  language: string,
  fieldBoostOverrides?: Record<string, number>,
): number {
  const avgFieldLength = ownProp(manifest.avgFieldLength, language) ?? {};
  let weightedTf = 0;
  for (const [field, fieldPosting] of Object.entries(posting.fields)) {
    const boost =
      fieldBoostOverrides?.[field] ?? manifest.fields[field]?.boost ?? 1.0;
    const avgLen = avgFieldLength[field] ?? fieldPosting.len;
    const lengthNorm = 1 - B + B * (fieldPosting.len / (avgLen || 1));
    weightedTf += (boost * fieldPosting.tf) / (lengthNorm || 1);
  }

  const docCount = ownProp(manifest.docCount, language) ?? 0;
  return idf(docCount, df) * (weightedTf / (weightedTf + K1));
}
